source/postprocess/tracers.cc

```diff
@@ -0,0 +1,406 @@
+#include <aspect/postprocess/tracers.h>
+#include <aspect/simulator_access.h>
+
+#include <aspect/particle/generator/interface.h>
+#include <aspect/particle/integrator/interface.h>
+#include <aspect/particle/interpolator/interface.h>
+#include <aspect/particle/property/interface.h>
+
+#include <boost/archive/text_oarchive.hpp>
+#include <boost/archive/text_iarchive.hpp>
+
+#include <stdio.h>
+#include <unistd.h>
+
+namespace aspect
+{
+  namespace Postprocess
+  {
+    template <int dim>
+    Tracers<dim>::Tracers ()
+      :
+      initialized(false),
+      next_data_output_time(std::numeric_limits<double>::quiet_NaN())
+    {}
+
+
+    template <int dim>
+    Tracers<dim>::~Tracers ()
+    {
+    }
+
+    template <int dim>
+    void
+    Tracers<dim>::initialize()
+    {
+
+    }
+
+    template <int dim>
+    const Particle::World<dim> &
+    Tracers<dim>::get_particle_world() const
+    {
+      return world;
+    }
+
+    template <int dim>
+    Particle::World<dim> &
+    Tracers<dim>::get_particle_world()
+    {
+      return world;
+    }
+
+    template <int dim>
+    std::pair<std::string,std::string>
+    Tracers<dim>::execute (TableHandler &statistics)
+    {
+      // Let the generator add the specified number of particles if we have not
+      // already done so
+      if (!initialized)
+        {
+          world.generate_particles();
+          world.initialize_particles();
+          initialized = true;
+          next_data_output_time = this->get_time();
+        }
+
+      // Advance the particles in the world to the current time
+      world.advance_timestep();
+
+      statistics.add_value("Number of advected particles",world.get_global_particle_count());
+
+      // If it's time to generate an output file and we created an output
+      // object, call the appropriate functions and reset the timer
+      if ((this->get_time() >= next_data_output_time) && output)
+        {
+          if (world.get_property_manager().need_update() == Particle::Property::update_output_step)
+            world.update_particles();
+
+          set_next_data_output_time (this->get_time());
+
+          const double output_time = (this->convert_output_to_years() ?
+                                      this->get_time() / year_in_seconds :
+                                      this->get_time());
+
+          const std::string data_file_name = output->output_particle_data(world.get_particles(),
+                                                                          world.get_property_manager().get_data_info(),
+                                                                          output_time);
+
+          // record the file base file name in the output file
+          statistics.add_value ("Particle file name",
+                                this->get_output_directory() + data_file_name);
+          return std::make_pair("Writing particle output: ", data_file_name);
+        }
+      return std::make_pair("","");
```

Comment:
How about returning something like
```
  std::make_pair ("Updating particles", Utilities::int_to_string (....) + " total particles");
```
so that the user gets to see something on the screen as well?